src/guiengine/event_handler.cpp

```diff
@@ -433,12 +433,12 @@ void EventHandler::sendNavigationEvent(const NavigationDirection nav, const int
         else if (nav == NAV_DOWN)
             propagation_state = widget_to_call->downPressed(playerID);
 
-        if (propagation_state == EVENT_LET)
-            sendEventToUser(widget_to_call, widget_to_call->m_properties[PROP_ID], playerID);
-
         if (propagation_state == EVENT_LET || propagation_state == EVENT_BLOCK_BUT_HANDLED)
             handled_by_widget = true;
 
+        if (propagation_state == EVENT_LET)
+            sendEventToUser(widget_to_call, widget_to_call->m_properties[PROP_ID], playerID);
+
```

Comment:
If I had inverted the values, the navigation would be entirely broken. And it isn't, obviously (most of this logic has been integrated with the previous PR).

First, you can have widgets inside widgets. In that case, a widget is not his own event handler. The loop body must always be done at least once (which is why I used a do... while in the first place), but it sometimes need to be done twice.

Next ; EVENT_LET/EVENT_BLOCK do not refer to the navigation, but to the propagation of the event to the main event handler (which is why sendEventToUser is called when it's EVENT_LET). For example, when the addons list return EVENT_LET ; it triggers the opening of the modal dialog with more info about the currently selected addon in the list. (That's the point of "EVENT_BLOCK_BUT_HANDLED" ; saying to the main navigation to not do anything without triggering other side effects.)

sendEventToUser doesn't return anything, but EVENT_LET is enough to know that no navigation event should be triggered.

All that said, I've built the code over the existing foundations, so the result is probably not as clear/simple as it could have been with new foundations, but rewriting existing code has its own downsides as it takes more time.